sys/kern/pty.c

```diff
@@ -0,0 +1,277 @@
+
+static devfs_node_t *pts_dir;
+
+typedef struct {
+  atomic_int pt_number; /* PTY number, if allocated. -1 means free. */
+  condvar_t pt_incv;    /* CV for readers */
+  condvar_t pt_outcv;   /* CV for writers */
+} pty_t;
+
+static pty_t pty_array[MAX_PTYS];
+
+static pty_t *pty_alloc(void) {
+  for (int i = 0; i < MAX_PTYS; i++) {
+    pty_t *pty = &pty_array[i];
+    if (atomic_compare_exchange_strong(&pty->pt_number, &(int){-1}, i + 1))
+      /* Got a free PTY. */
+      return pty;
+  }
+  return NULL;
+}
+
+static void pty_free(pty_t *pty) {
+  assert(pty->pt_number >= 0);
+  pty->pt_number = -1;
+}
+
+static int pty_read(file_t *f, uio_t *uio) {
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+  int error;
+  size_t start_resid = uio->uio_resid;
+
+  if (uio->uio_resid == 0)
+    return 0;
+
+  SCOPED_MTX_LOCK(&tty->t_lock);
+
+  /* Wait until there is at least one byte of data. */
+  while (ringbuf_empty(&tty->t_outq)) {
+    /* Don't wait for data if slave device isn't opened. */
+    if (!tty_opened(tty))
+      return 0;
+    if (cv_wait_intr(&pty->pt_incv, &tty->t_lock))
+      return ERESTARTSYS;
+  }
+
+  /* Data is available: transfer as much as we can. */
+  error = ringbuf_read(&tty->t_outq, uio);
+  tty_getc_done(tty);
+
+  /* Don't report errors on partial reads. */
+  if (start_resid > uio->uio_resid)
+    error = 0;
+
+  return error;
+}
+
+/* Write at single character to the master side of a pseudoterminal,
+ * i.e. to the input queue of the slave tty.
+ * If the input queue is full, sleep only if the slave tty has users. */
+static int pty_putc_sleep(tty_t *tty, pty_t *pty, uint8_t c) {
+  while (!tty_input(tty, c)) {
+    if (!tty_opened(tty))
+      return EIO;
+    if (cv_wait_intr(&pty->pt_outcv, &tty->t_lock))
+      return ERESTARTSYS;
+  }
+  return 0;
+}
+
+static int pty_write(file_t *f, uio_t *uio) {
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+  int error = 0;
+  uint8_t c;
+
+  if (uio->uio_resid == 0)
+    return 0;
+
+  size_t start_resid = uio->uio_resid;
+
+  SCOPED_MTX_LOCK(&tty->t_lock);
+
+  while (uio->uio_resid > 0) {
+    if ((error = uiomove(&c, 1, uio)))
+      break;
+    if ((error = pty_putc_sleep(tty, pty, c))) {
+      /* Undo the last uiomove(). */
+      uio->uio_resid++;
+      break;
+    }
+  }
+
+  /* Don't report errors on partial writes. */
+  if (start_resid > uio->uio_resid)
+    error = 0;
+
+  return error;
+}
+
+static int pty_close(file_t *f) {
+  /* PTY master devices can't be accessed via the filesystem, so here we know
+   * that the device won't ever be used again. */
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+
+  mtx_lock(&tty->t_lock);
+  tty_detach_driver(tty); /* Releases t_lock. */
+  pty_free(pty);
+  return 0;
+}
+
+static int pty_seek(file_t *f, off_t offset, int whence, off_t *newoffp) {
+  return 0;
+}
+
+static int pty_stat(file_t *f, stat_t *sb) {
+  memset(sb, 0, sizeof(stat_t));
+  sb->st_mode = S_IFCHR;
```

Comment:
If I perform `stat` on pseudoterminal device it reports:

```
  File: /dev/pts/5
  Size: 0         	Blocks: 0          IO Block: 1024   character special file
Device: 13h/19d	Inode: 8           Links: 1     Device type: 88,5
Access: (0620/crw--w----)  Uid: ( 1001/   cahir)   Gid: (    5/     tty)
Access: 2021-01-28 08:03:36.135880424 +0100
Modify: 2021-01-28 08:03:36.135880424 +0100
Change: 2021-01-27 00:32:48.135880424 +0100
```

Aren't we missing something here?

Reply:
`/dev/pts/5` is a slave tty, `pty_stat()` implements `FOP_STAT()` for the master side. I agree it's somewhat bare-bones... FreeBSD appears to take most attributes from the slave device ([link](https://grok.dragonflybsd.org/xref/freebsd/sys/kern/tty_pts.c?r=77fb93e1#538)), so I will do that too.